src/arena.c

```diff
@@ -292,34 +292,51 @@ arena_slab_reg_alloc_batch(edata_t *slab, const bin_info_t *bin_info,
 
 static void
 arena_large_malloc_stats_update(tsdn_t *tsdn, arena_t *arena, size_t usize) {
-	szind_t index, hindex;
-
 	cassert(config_stats);
 
+	szind_t index = sz_size2index(usize);
 	if (usize < SC_LARGE_MINCLASS) {
-		usize = SC_LARGE_MINCLASS;
+		assert(index < SC_NBINS);
+		bin_t *bin = &arena->bins[index];
+		malloc_mutex_lock(tsdn, &bin->lock);
+		bin->stats.nmalloc++;
+		malloc_mutex_unlock(tsdn, &bin->lock);
+	} else {
+		assert(index >= SC_NBINS);
+		szind_t hindex = index - SC_NBINS;
+		LOCKEDINT_MTX_LOCK(tsdn, arena->stats.mtx);
+		locked_inc_u64(tsdn, LOCKEDINT_MTX(arena->stats.mtx),
+			&arena->stats.lstats[hindex].nmalloc, 1);
+		LOCKEDINT_MTX_UNLOCK(tsdn, arena->stats.mtx);
 	}
-	index = sz_size2index(usize);
-	hindex = (index >= SC_NBINS) ? index - SC_NBINS : 0;
-
-	locked_inc_u64(tsdn, LOCKEDINT_MTX(arena->stats.mtx),
-	    &arena->stats.lstats[hindex].nmalloc, 1);
 }
 
 static void
 arena_large_dalloc_stats_update(tsdn_t *tsdn, arena_t *arena, size_t usize) {
-	szind_t index, hindex;
-
 	cassert(config_stats);
 
-	if (usize < SC_LARGE_MINCLASS) {
-		usize = SC_LARGE_MINCLASS;
+	/*
+	 * The key difference between this function and
+	 * `arena_large_malloc_stats_update` above is that here `usize` does
+	 * _not_ already include size promotion for sampled small allocations.
+	 */
```

Comment:
Don't we already restored the usize by this point? It's done here: https://github.com/jemalloc/jemalloc/blob/dev/src/arena.c#L1269

which is before the actual dalloc logic

Reply:
Exactly, `usize` is no longer bumped here, but we want to attribute the stats to the size-class which was actually used to service this allocation, which _was_ bumped at the time.

Reply:
Not following. I meant demote already happened before the dalloc, so the extent usize is already the bumped page sizes. It looks we can assert(usize % PAGE == 0) here.

Reply:
Okay, I see the real issue now. I originally assumed I had gotten something wrong in general due to unit-test failures. But I think the unit-test failure is actually an edge-case. Namely, in the `arena_reset_prof` test, the deallocation that gets initiated [on this line](https://github.com/jemalloc/jemalloc/blob/3e82f357bb218194df5ba1acee39cd6a7d6fe6f6/src/arena.c#L700) is tripping your proposed assertion.

Reply:
Oh in that case, it doesn't do `arena_dalloc_promoted`, is that right?

Reply:
Ah actually I had just made a mistake. The other feedback I left to your response above needed to be incorporated (so we use `bumped_usize` when appropriate), and I was also missing parentheses around the assignment of `bumped_usize` so it was getting set to 0 or 1 instead of the actual value. Fixed, everything is passing now.

Reply:
I think you probably caught an issue with the reset operation. Can you check if `arena_dalloc_promoted` is missed, if prof is enabled while doing reset? https://github.com/jemalloc/jemalloc/blob/dev/include/jemalloc/internal/arena_inlines_b.h#L302

Reply:
Stepping through in a debugger reveals that `arena_prof_promote` is called many times, but `arena_prof_demote` is called many fewer. I can produce this on `dev` without the changes from this PR. So you're right, this _is_ an issue with reset. I'll look into solving it now.

Reply:
Opened #2496 to fix this.

Reply:
Rebased now that #2496 has been merged. Eliminated the unnecessary bumping logic here now that has been fixed.